src/front/glsl/builtins.rs

```diff
@@ -1429,88 +1393,127 @@ impl MacroCall {
                 ctx.samplers.insert(args[0], args[1]);
                 Ok(args[0])
             }
-            MacroCall::Texture | MacroCall::TextureLod => {
-                let comps = parser.coordinate_components(ctx, args[0], args[1], meta, body)?;
-                let level = match *self {
-                    MacroCall::Texture => args
-                        .get(2)
-                        .copied()
-                        .map_or(SampleLevel::Auto, SampleLevel::Bias),
-                    MacroCall::TextureLod => SampleLevel::Exact(args[2]),
-                    _ => unreachable!(),
-                };
-                texture_call(ctx, args[0], level, comps, body, meta)
-            }
-            MacroCall::TextureGrad(shadow) => {
-                let comps = parser.coordinate_components(ctx, args[0], args[1], meta, body)?;
-                let level = match shadow {
-                    true => {
-                        log::debug!(
-                            "Assuming gradients {:?} and {:?} are not greater than 1",
-                            args[2],
-                            args[3]
-                        );
-                        SampleLevel::Zero
-                    }
-                    false => SampleLevel::Gradient {
-                        x: args[2],
-                        y: args[3],
-                    },
-                };
-                texture_call(ctx, args[0], level, comps, body, meta)
-            }
-            MacroCall::TextureProj => {
-                let level = args
-                    .get(2)
-                    .copied()
-                    .map_or(SampleLevel::Auto, SampleLevel::Bias);
-                let size = match *parser.resolve_type(ctx, args[1], meta)? {
-                    TypeInner::Vector { size, .. } => size,
-                    _ => {
-                        return Err(Error {
-                            kind: ErrorKind::SemanticError("Bad call to textureProj".into()),
-                            meta,
-                        })
-                    }
-                };
-                let base = args[1];
-                let mut right = ctx.add_expression(
-                    Expression::AccessIndex {
-                        base,
-                        index: size as u32 - 1,
-                    },
-                    SourceMetadata::none(),
-                    body,
-                );
-                let left = if let VectorSize::Bi = size {
-                    ctx.add_expression(
-                        Expression::AccessIndex { base, index: 0 },
+            MacroCall::Texture { proj, offset, shadow, level_type } => {
+                let mut coords = args[1];
+
+                if proj {
+                    let size = match *parser.resolve_type(ctx, args[1], meta)? {
+                        TypeInner::Vector { size, .. } => size,
+                        _ => {
```

Comment:
nit: this is unreachable by design, maybe use `VectorSize::Bi` as default and add a comment saying it's unreachable as the functions are type checked

Reply:
This code came straight from the older proj code, I believe. Should I just use `unreachable!()`?

Reply:
Decided to use unreachable!()